### DISCUSSION TITLE
The SSL connection could not be established, see inner exception.

### DISCUSSION THREAD

**Discussion by xbaha** [betalgo/openai#183] [Announcements]
**Describe the bug**
The SSL connection could not be established, see inner exception.
SocketException: An existing connection was forcibly closed by the remote host.

**Your code piece**
```
        public async Task GenerateResponse2()
        {
            var openAiService = new OpenAIService(new OpenAiOptions()
            {
                ApiKey = apiKey
            });

            var completionResult = await openAiService.ChatCompletion.CreateCompletion(new ChatCompletionCreateRequest
            {
                Messages = new List<ChatMessage>
                {
                    ChatMessage.FromSystem("You are a helpful assistant."),
                    ChatMessage.FromUser("Who won the world series in 2020?"),
                    ChatMessage.FromAssistant("The Los Angeles Dodgers won the World Series in 2020."),
                    ChatMessage.FromUser("Where was it played?")
                },
                Model = Models.ChatGpt3_5Turbo,
                MaxTokens = 50//optional
            });
            if (completionResult.Successful)
            {
                Console.WriteLine(completionResult.Choices.First().Message.Content);
            }
        }
```

**Comment by kayhantolga** [comment]
If you need to use a self-signed SSL certificate you may want to pass your own HTTP client to the library. There was a topic in the discussion about this.

**Reply by xbaha** [reply]
What do you mean?? I didn't use any self-signed SSL, I just installed the nuget package, insert the demo function, and got this error!

**Comment by pdcruze** [comment]
What OS are you running this on?
OpenAI API support TLS1.3 only.  .NET uses the native windows SSL/TLS functionality. TLS1.3 is only supported by Windows 10 and newer I  believe.

**Reply by xbaha** [reply]
I am using Windows 11

**Comment by xbaha** [comment]
Update: I tried the same thing on Windows 10 and it worked ok, does this mean windows 11 uses > TLS1.3 version? could you force the NuGet package to use this TLS version?

**Comment by George-Smy** [comment]
I am seeing the same issue with a recently patched Windows Server 2012 R2 server. I've tried adding a standard line in VB.Net to force use of "TLS13":

`System.Net.ServicePointManager.SecurityProtocol = System.Net.SecurityProtocolType.Tls13`

just before OpenAI.GPT3.Managers.OpenAIService but I'm seeing the 'HandshakeFailure' failure message. Are there any other options besides updating the server OS which may not be possible. I've been developing on a Window 10 system for the past few weeks without issue and this is the first time the app has been off my system for testing.

Thanks!

**Comment by xbaha** [comment]
The problem is not with this repo, as I tried direct HttpClient
```
        public async Task<ChatCompletion> GenerateResponse(List<Message> messages) {
            string endpoint = "https://api.openai.com/v1/chat/completions";

            HttpClientHandler clientHandler = new HttpClientHandler
            {
                ClientCertificateOptions = ClientCertificateOption.Manual,
                ServerCertificateCustomValidationCallback = (sender, cert, chain, sslPolicyErrors) => true,
                SslProtocols = System.Security.Authentication.SslProtocols.Tls13,
            };

            using (HttpClient client = new HttpClient(clientHandler)) {
                client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
                client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

                var requestBody = new {
                    model = "gpt-3.5-turbo",
                    messages = messages
                };

                var json = JsonConvert.SerializeObject(requestBody);
                var content = new StringContent(json, Encoding.UTF8, "application/json");

                var response = await client.PostAsync(endpoint, content);
                var responseString = await response.Content.ReadAsStringAsync();

                if (!response.IsSuccessStatusCode) {
                    throw new Exception($"OpenAI API call failed with status code {response.StatusCode}: {responseString}");
                }

                var responseData = JsonConvert.DeserializeObject<ChatCompletion>(responseString);
                return responseData;
            }
        }
```
The code above worked on Windows10, Now i am using Windows 11, 
Someone mentioned that the ChatGPT server uses TLS1.3, I have been trying to add TLS1.3 to windows 11 but failed to make the call.

I tried this code to force TLS1.3:
```
            HttpClientHandler clientHandler = new HttpClientHandler
            {
                ClientCertificateOptions = ClientCertificateOption.Manual,
                ServerCertificateCustomValidationCallback = (sender, cert, chain, sslPolicyErrors) => true,
                SslProtocols = System.Security.Authentication.SslProtocols.Tls13,
            };
```

and tried adding to the registry (https://www.asustor.com/en-gb/knowledge/detail/?id=&group_id=1011)

see if any method works for you and update back, please.

**Reply by Frogley** [reply]
My code can run normally in both Windows 10 and Windows 11. Sorry I haven't encountered the same situation as you.

**Comment by leohilbert** [comment]
I'm having the same situation. Did somebody find a solution? Such a weird issue
